Option for downloading tissue image for spatial visium dataset

hspitzer opened the pull request:
Following discussions with @giovp I've extended the `scanpy.datasets.visium_sge` function to optionally return a path to the high-resolution tissue image also available in the visium Spatial Transcriptomics datasets.
This makes it easy to leverage `scanpy.datasets` to fully explore visium datasets.

hspitzer committed:
allow downloading tissue image from visium dataset

ivirshup commented:
Why not store the path to the image inside `uns` with the other image metadata?

ivirshup commented on scanpy/tests/test_datasets.py:
```diff
@@ -98,6 +98,12 @@ def test_visium_datasets(tmp_dataset_dir, tmpdir):
+    mbrain, mbrain_img = sc.datasets.visium_sge(
+        "V1_Adult_Mouse_Brain", return_image_path=True
+    )
+    assert mbrain_img == sc.settings.datasetdir / "V1_Adult_Mouse_Brain" / "image.tif"
```
Could this also make sure the file exists, maybe even is an image?

hspitzer replied:
I'm now checking with `file` whether the file is a tiff

hspitzer commented:
> Why not store the path to the image inside `uns` with the other image metadata?

yes, good idea - I'll do that instead of returning the path!

hspitzer committed:
add image path to adata.uns instead of returning

hspitzer committed:
store image path with correct sample_id

hspitzer commented:
I've updated the code to store the image path in `adata.uns['spatial'][sample_id]['tif_image_path']`. The test now also checks whether the image file exists. 

I can also test whether the image file is a valid tiff image, but for this we'd need to add a tiff reading library like `pillow` as a test dependency to scanpy. If that is ok, I'll update the PR with an additional test.

giovp commented:
thanks a lot @hspitzer !
I think checking for image is indeed problematic, maybe we can drop it if it requests additional packages.
re: image path. I would store it in the metadata dict:
```python
adata.uns["spatial"]["V1_Adult_Mouse_Brain"]["metadata"]
>>> {'chemistry_description': "Spatial 3' v1",
>>> 'software_version': 'spaceranger-1.0.0'}
```
with some name like key=image_name and value=path ? or something like that.

hspitzer commented:
> with some name like key=image_name and value=path ? or something like that.

sure! What do you mean with image_name? As downloaded from the 10x website, the tiff image always has the same filename ("image.tif"). Or do you mean the string "image_name" as key?

giovp commented:
good point, maybe the img format? like `'tif_image':''[PATH]`
so that if it's jpeg than it's visible?

ivirshup commented:
Should this argument be added to `sc.read_visium` as well? That is, add a path to the tiff if it would be in a standardized place?

ivirshup commented:
Also, I don't think I would mind `Pillow` too much as a test dependency (it doesn't have a ton of dependencies, right?)

A more lightweight solution would be to call [`file`](https://en.wikipedia.org/wiki/File_(command)) on the path, which should be able to recognize it as a `tiff`. I think this would only look at the head of the file though, and wouldn't check if it was corrupted/ didn't finish downloading.

ivirshup commented on scanpy/datasets/_datasets.py:
```diff
@@ -386,7 +399,9 @@ def visium_sge(
     sample_id
         The ID of the data sample in 10x’s spatial database.
-
+    add_image_path
```
I think this could have a better name. Maybe `download_tiff` or `include_tiff`?

hspitzer replied:
yes, I'll change it to `download_tif`. Or is there a preference for using "tiff" over "tif"?

giovp replied:
since image can be non-tif, I would choose a different name

hspitzer replied:
maybe `download_tissue_image`?

giovp commented:
> Should this argument be added to `sc.read_visium` as well? That is, add a path to the tiff if it would be in a standardized place?

yes that's a very good point actually. It could even make it easier for us to initialize the image container maybe?

hspitzer commented:
> > Should this argument be added to `sc.read_visium` as well? That is, add a path to the tiff if it would be in a standardized place?
> 
> yes that's a very good point actually. It could even make it easier for us to initialize the image container maybe?

I thought about that already - standardised places for the tiff would be `sample_id"_image.tif` and `image.tif`. I don't think we need to add an extra argument to `sc.read_visium`, couldn't it automatically add the path if it exists?

hspitzer commented:
> Also, I don't think I would mind `Pillow` too much as a test dependency (it doesn't have a ton of dependencies, right?)
> 
> A more lightweight solution would be to call [`file`](https://en.wikipedia.org/wiki/File_(command)) on the path, which should be able to recognize it as a `tiff`. I think this would only look at the head of the file though, and wouldn't check if it was corrupted/ didn't finish downloading.

I was not aware of `file`. I think it might be a good solution! `readwrite._download` should make sure that downloads are not incomplete, so just reading the header might be enough

giovp commented:
> > > Should this argument be added to `sc.read_visium` as well? That is, add a path to the tiff if it would be in a standardized place?
> > 
> > 
> > yes that's a very good point actually. It could even make it easier for us to initialize the image container maybe?
> 
> I thought about that already - standardised places for the tiff would be `sample_id"_image.tif` and `image.tif`. I don't think we need to add an extra argument to `sc.read_visium`, couldn't it automatically add the path if it exists?

good point, agree probably best solution, however one thing it's not given is the format of the file (could be tif or jpg, but potentially some else, depending on the lab). But an automatic search would be best, just not sure how extensive we can make it

hspitzer commented:
> > > > Should this argument be added to `sc.read_visium` as well? That is, add a path to the tiff if it would be in a standardized place?
> > > 
> > > 
> > > yes that's a very good point actually. It could even make it easier for us to initialize the image container maybe?
> > 
> > 
> > I thought about that already - standardised places for the tiff would be `sample_id"_image.tif` and `image.tif`. I don't think we need to add an extra argument to `sc.read_visium`, couldn't it automatically add the path if it exists?
> 
> good point, agree probably best solution, however one thing it's not given is the format of the file (could be tif or jpg, but potentially some else, depending on the lab). But an automatic search would be best, just not sure how extensive we can make it

do you have an example where the image is a jpg? I can account for both tif and jpg, but I'm wondering if its worth it.

giovp commented:
> > > > > Should this argument be added to `sc.read_visium` as well? That is, add a path to the tiff if it would be in a standardized place?
> > > > 
> > > > 
> > > > yes that's a very good point actually. It could even make it easier for us to initialize the image container maybe?
> > > 
> > > 
> > > I thought about that already - standardised places for the tiff would be `sample_id"_image.tif` and `image.tif`. I don't think we need to add an extra argument to `sc.read_visium`, couldn't it automatically add the path if it exists?
> > 
> > 
> > good point, agree probably best solution, however one thing it's not given is the format of the file (could be tif or jpg, but potentially some else, depending on the lab). But an automatic search would be best, just not sure how extensive we can make it
> 
> do you have an example where the image is a jpg? I can account for both tif and jpg, but I'm wondering if its worth it.

collaborators that generated visium data sent me jpg

hspitzer committed:
modify read_visium to add path to tif image to adata

hspitzer committed:
remove print in test

hspitzer commented:
I've update the code to 
- test that the file is actually a tiff image
- automatically add the path to the image to `adata.uns['spatial'][library_id]['metadata']['tissue_image_path']`. It's looking for a tiff or jpeg file with the name `"image"` or `library_id"_image"`. This should cover most cases hopefully?

ivirshup commented:
I think it would be fine to only cover the case of what `space ranger` actually outputs. I was thinking there could be an argument where the user manually passes an alternate path. This could be useful for cases where they've processed the image themselves some modifications to the image.

giovp commented:
> I think it would be fine to only cover the case of what `space ranger` actually outputs. I was thinking there could be an argument where the user manually passes an alternate path. This could be useful for cases where they've processed the image themselves some modifications to the image.

space ranger doesn't output this image, as it's taken as input to assign spots and get scalefactors and metadata. This type of image is in the same folder just for chance in the 10x genomics dataset. 
In the `read_visium` function I would simply add an argument to pass the path of the image, and basically just assign it to the `adata.uns` metadata. Otherwise just assign None. THis way it's consistent for the spatial tool whichlater uses it in the image container.
It's also convenient to add it as argument so that `read_visium` could just be passed in that same way as it is now in `datasets.visium_sge`

hspitzer committed:
rename argument, add tissue_image_path

hspitzer commented:
I've made a small update to add a `tissue_image_path` parameter to `read_visium` that you can use to specify the tissue image path to put in `adata.uns`

hspitzer committed:
fix formatting

ivirshup commented:
/AzurePipelines run

ivirshup commented on scanpy/readwrite.py:
```diff
@@ -418,6 +422,26 @@ def read_visium(
+                    break
+        # put image path in uns
+        else:
+            tissue_image_path = Path(tissue_image_path).resolve()
```
Is it very important that you guess where the image is? I would rather this not happen, since it could easily read the wrong thing, or not find an image a user would want to pick up. I'd rather the user had to be explicit about this.

hspitzer replied:
From a user perspective its more convenient if the path is assumed. Not that this is also the case for the png images (assumed to be in `spatial/tissue_hires_image.png`). This is the case for the standardised datasets you can download from visium but many of course be different for datasets that you create yourself. 
I do understand your point though, guessing is rarely a very good idea.

ivirshup replied:
I think there are two issues with this:

* If it's wrong, it's not obvious until later in the analysis. If there are other images in the directory, for example: low-res thumbnails or plots from the user's pipeline could get picked up by accident.
* As it stands right now, if a user ever passed `"include_tiff"` (or whatever we end up calling it), it will always be included on subsequent calls. I think it's bad for there to be state like that in these functions.

hspitzer replied:
ok, sure. Lets leave it at having the user passing `source_image_path` to `read_visium` when they want to include this information in the adata.

ivirshup commented on scanpy/datasets/_datasets.py:
```diff
@@ -377,6 +389,7 @@ def visium_sge(
         'Targeted_Visium_Human_ColorectalCancer_GeneSignature',
         'Parent_Visium_Human_ColorectalCancer',
     ] = 'V1_Breast_Cancer_Block_A_Section_1',
+    download_tissue_image: bool = False,
```
Instead of `tissue_image`, could this be something like `tiff`, `hires_tiff`, or `raw_tiff`?

I don't like `tissue_image`. The jpegs we load by default could also be called `tissue_image`s, so this seems a little unclear to me.

hspitzer replied:
sure, I can call this flag `download_hires_tiff`. This is only for downloading images from 10x, and they only provide hires tiff images.
@giovp made the point earlier that in general, the hires_tiff image could also be a jpg, so I would still call the entry in `adata.uns` something like `tissue_image_path` or `hires_image_path`.

giovp replied:
agree, argument of download could have `tiff` but argument in `uns` should not. I would also drop hires cause there is already a hires in `adata.uns["spatial"]`. Maybe `full_tissue_image`  or something like that? don't know sorry

ivirshup replied:
What uses these images in `squidpy`? In particular what uses these images that wouldn't work with the `jpeg`s?

giovp replied:
in `squidpy` this image is used by the image container. Don't understand the `jpeg` question sorry

ivirshup replied:
Ha, for `jpeg` I meant the `tissue_hires_image.png` and `tissue_lowres_image.png`.

ivirshup replied:
I think the implication is that this is the `raw`, `full`. or `source` image that you want to do processing on.

I would agree that `hires` would also be a good descriptor, but that's already used by the `png`.

You would be in a better position to comment, but I like `source_image_path` and `full_image_path`. Both imply that they are the complete data, which is what you'll want to do processing on.

giovp replied:
ah ok, so 
> what uses these images that wouldn't work with the jpegs [pngs]?

basically everything that works on imag in squidpy, in particular segmentation, feature extraction and visualization with napari

> You would be in a better position to comment, but I like source_image_path and full_image_path. Both imply that they are the complete data, which is what you'll want to do processing on.

Also agree, actually `source_image_path` is nice as that image is indeed the source image for spaceranger to do the processing

ivirshup reviewed (changes requested):
Sorry for the late review! A few comments about naming conventions and guessing filenames.

> In the read_visium function I would simply add an argument to pass the path of the image, and basically just assign it to the adata.uns metadata. Otherwise just assign None. THis way it's consistent for the spatial tool whichlater uses it in the image container.

I like this approach.

hspitzer committed:
rename arguments, remove guessing tiff file name

hspitzer commented:
Updated the PR:
- renamed argument in `visium_sge` to `include_hires_tiff`
- renamed argument in `read_visium` to `source_image_path` and removed guessing of image file if `source_image_path` is None

ivirshup commented on scanpy/readwrite.py:
```diff
@@ -418,6 +421,14 @@ def read_visium(
+            source_image_path = str(Path(source_image_path).resolve())
+        adata.uns["spatial"][library_id]["metadata"]["source_image_path"] = str(
+            source_image_path
+        )
```
```suggestion
            adata.uns["spatial"][library_id]["metadata"]["source_image_path"] = str(
                source_image_path
            )
```

Right?

hspitzer replied:
yes, thanks, totally missed that!

ivirshup commented on scanpy/tests/test_datasets.py:
```diff
@@ -7,6 +7,8 @@
 import pytest
 from pathlib import Path
 from anndata.tests.helpers import assert_adata_equal
+import os
```
```suggestion
```

This doesn't get used, right?

hspitzer replied:
no, not anymore

ivirshup commented on scanpy/readwrite.py:
```diff
@@ -316,6 +317,8 @@ def read_visium(
     library_id
         Identifier for the visium library. Can be modified when concatenating multiple adata objects.
+    source_image_path
+        Path to the high-resolution tissue image. Will be saved in `uns`.
```
```suggestion
    source_image_path
        Path to the high-resolution tissue image. Path will be included in
        `.uns["spatial"][sample_id]["metadata"]["source_image_path"]`.
```

ivirshup reviewed (changes requested):
Minor changes, then lgtm.

Please also add a release note!

ivirshup commented on scanpy/datasets/_datasets.py:
```diff
@@ -386,7 +399,9 @@ def visium_sge(
-
+    include_hires_tiff
+        Download and include the high-resolution tissue image (tiff) in `adata.uns["spatial"][sample_id]["metadata"]["source_image_path"]`.
+        
```
```suggestion
    include_hires_tiff
        Download and include the high-resolution tissue image (tiff) in `adata.uns["spatial"][sample_id]["metadata"]["source_image_path"]`.

```

ivirshup commented on scanpy/datasets/_datasets.py:
```diff
@@ -377,6 +389,7 @@ def visium_sge(
         'Targeted_Visium_Human_ColorectalCancer_GeneSignature',
         'Parent_Visium_Human_ColorectalCancer',
     ] = 'V1_Breast_Cancer_Block_A_Section_1',
+    include_hires_tiff: bool = False,
```
```suggestion
    ] = 'V1_Breast_Cancer_Block_A_Section_1',
    *,
    include_hires_tiff: bool = False,
```

I think it makes sense for this argument to be keyword only.

hspitzer committed:
Apply suggestions from code review

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

hspitzer commented:
> Please also add a release note!

sorry for the naive question, but how do I add a release note? As a comment in the PR? As a commit?

giovp commented:
just mention this PR with brief description and your name here: https://github.com/theislab/scanpy/blob/master/docs/release-latest.rst

nice! only one PR to go, thank you @hspitzer @ivirshup !

hspitzer committed:
add release note

hspitzer commented:
> just mention this PR with brief description and your name here: https://github.com/theislab/scanpy/blob/master/docs/release-latest.rst

ok, done. I put it in the "on master" section

hspitzer committed:
Merge branch 'master' into feature_download_visium_image

ivirshup commented on scanpy/readwrite.py:
```diff
@@ -316,6 +317,9 @@ def read_visium(
         Identifier for the visium library. Can be modified when concatenating multiple adata objects.
+    source_image_path
+        Path to the high-resolution tissue image. Path will be included in
+        `.uns["spatial"][sample_id]["metadata"]["source_image_path"]`.
```
```suggestion
        `.uns["spatial"][library_id]["metadata"]["source_image_path"]`.
```

ivirshup committed:
Move change note to "features" section

ivirshup committed:
Typo, sample_id -> library_id

ivirshup commented:
Thanks for going through this with me!
